services/web/server/src/simcore_service_webserver/computation_comp_tasks_listening_task.py

```diff
@@ -0,0 +1,119 @@
+"""this module creates a background task that monitors changes in the database.
+First a procedure is registered in postgres that gets triggered whenever the outputs
+of a record in comp_task table is changed.
+"""
+
+import asyncio
+import json
+import logging
+
+from aiohttp import web
+from aiopg.sa import Engine
+from sqlalchemy.sql import select
+
+from servicelib.application_keys import APP_DB_ENGINE_KEY
+
+from .computation_config import APP_COMP_TASKS_LISTENING_KEY
+from .projects import projects_api
+from .projects.projects_models import projects, user_to_projects
+
+log = logging.getLogger(__name__)
+
+DB_PROCEDURE_NAME: str = "notify_comp_tasks_changed"
+DB_TRIGGER_NAME: str = f"{DB_PROCEDURE_NAME}_event"
+DB_CHANNEL_NAME: str = "comp_tasks_output_events"
+async def register_trigger_function(app: web.Application):
+    db_engine: Engine = app[APP_DB_ENGINE_KEY]
+    # NOTE: an example was found in https://citizen428.net/blog/asynchronous-notifications-in-postgres/
+    notification_fct_query = f"""    
+    CREATE OR REPLACE FUNCTION {DB_PROCEDURE_NAME}() RETURNS TRIGGER AS $$
+        DECLARE
+            record RECORD;
+            payload JSON;
+        BEGIN
+            IF (TG_OP = 'DELETE') THEN
+                record = OLD;
+            ELSE
+                record = NEW;
+            END IF;
+
+            payload = json_build_object('table', TG_TABLE_NAME,
+                                        'action', TG_OP,
+                                        'data', row_to_json(record));
+
+            PERFORM pg_notify('{DB_CHANNEL_NAME}', payload::text);
+
+            RETURN NULL;
+        END;
+    $$ LANGUAGE plpgsql;
+    """
+
+    trigger_registration_query = f"""
+    CREATE TRIGGER {DB_TRIGGER_NAME}
+    AFTER UPDATE OF outputs ON comp_tasks
+        FOR EACH ROW
+        WHEN (OLD.outputs::jsonb IS DISTINCT FROM NEW.outputs::jsonb)
+        EXECUTE PROCEDURE {DB_PROCEDURE_NAME}();
+    """
+    async with db_engine.acquire() as conn:
+        await conn.execute(notification_fct_query)
+        await conn.execute(trigger_registration_query)
```

Comment:
use [``conn.begin()``](https://aiopg.readthedocs.io/en/stable/sa.html#aiopg.sa.SAConnection.begin) to rollback transactions

see [examples](https://aiopg.readthedocs.io/en/stable/examples.html#simple-transaction-in-sqlalchemy)

Reply:
done